### PR TITLE
[SPARK-19892][MLlib] Implement findAnalogies method for Word2VecModel

### PR DESCRIPTION
**Pull request by benradford**
## What changes were proposed in this pull request?

Added findAnalogies method to Word2VecModel for performing vector-algebra-based queries (e.g. King + Woman - Man).

## How was this patch tested?

Followed the contributor's guide for Spark and ran the run-tests. Compiled and tested functionality in spark-shell.

This is an original work that I license to the project under the project's open source license.

### CODE REVIEW

**Comment by benradford** [comment]
ok to test
Jenkins, add to whitelist

File: R/pkg/DESCRIPTION [hunk 1]
```diff
@@ -54,5 +54,5 @@ Collate:
     'types.R'
     'utils.R'
     'window.R'
-RoxygenNote: 5.0.1
+RoxygenNote: 6.0.1
```
Comment (felixcheung) [inline R/pkg/DESCRIPTION]: please revert this.

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/DESCRIPTION (hunk 1 above)
Comment (srowen) [inline R/pkg/DESCRIPTION]: This happens automatically. Ideally the build doesn't modify source files -- do you know any way to make it not do this?

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 2]
```diff
@@ -610,6 +610,71 @@ class Word2VecModel private[spark] (
...
   /**
+   * Find words similar to the words supplied to 'positive' and dissimilar
+   * to the words supplied to 'negative'.
+   * @param positive array of words similar to the results list
+   * @param negative array of words dissimilar to the results list
+   * @param num number of synonyms to find
+   * @return array of (word, cosineSimilarity)
+   */
+  def findAnalogies(positive: Array[String] = Array(),
+                            negative: Array[String] = Array(),
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Nit: indentation.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 3]
```diff
@@ -610,6 +610,71 @@ class Word2VecModel private[spark] (
...
+    var positiveVectors = Array[Array[Double]]()
+    var negativeVectors = Array[Array[Double]]()
+
+    for(pp <- positive)
+      positiveVectors :+= transform(pp).toArray
+    for(nn <- negative)
+      negativeVectors :+= transform(nn).toArray
+    // Normalize positive and negative vectors before summation
+    positiveVectors = if (positiveVectors.size > 0) {
+      positiveVectors.map(x => {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: There is a fair bit of duplicated code that could be factored into methods.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 4]
```diff
@@ -610,6 +610,71 @@ class Word2VecModel private[spark] (
...
+        val sumsqr = x.map(y => y * y).reduce((a, b) => a + b)
+        x.map(y => y / math.pow(sumsqr, .5))
+      })
+    } else {
+      Array(Array.fill(vectorSize)(0.0))
+    }
+    negativeVectors = if (negativeVectors.size > 0) {
+      negativeVectors.map(x => {
+        val sumsqr = x.map(y => y * y).reduce((a, b) => a + b)
+        x.map(y => y / math.pow(sumsqr, .5))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: You just mean sqrt right? pow 0.5 is less readable

**Review by srowen — COMMENTED** [review]
I'm not sure if this adds enough value to be in the core implementation. Any caller can do this. Is it common enough that it needs to be a convenience method in Spark?

**Comment by benradford** [comment]
@srowen I find this functionality useful and thought others might as well. It is a common use-case for word2vec and is suggested as a method for validating good model fit by Mikolov et al. I understand if you prefer not to include it in the core implementation, too. I did my best to address your suggestions in my latest commit.

File: R/pkg/DESCRIPTION (hunk 1 above)
Comment (felixcheung) [inline R/pkg/DESCRIPTION]: It does it automatically when someone having a newer roxygen2 installed is running the R build.
From looking at that code, there isn't an option to disable the behavior of updating this string automatically. And, I don't think there is a way to fix the version either, other than mucking with R packages that are installed.

**Review by benradford — COMMENTED** [review]
Corrected roxygen

**Comment by benradford** [comment]
@srowen @felixcheung: I believe all concerns have been addressed. Please let me know if there are any remaining issues.

**Review by benradford — COMMENTED** [review]
fixed.

**Pull request closed without merging**